### PR TITLE
More accurate pinhole camera calibration with imperfect planar target

### PR DESCRIPTION
**Pull request by xoox**
**Merge with extra**: opencv/opencv_extra#528
### Problem to be solved

Checkerboard patterns on paper using off-the-shelf printers are the most
convenient calibration targets. And these printed patterns stuck on
roughly planar surfaces are the mostly used calibration structure world
wide. But in most cases printed patterns are inaccurate and the sticking
surfaces are not perfectly planar.

chessboard

For example, the square size of the above chessboard is set to 15mm
in the PDF file. The printed paper was pasted to a glass panel which was
not so perfect planar structure. We measured the distances between
combinations of p1, p2, p3 and p4 with a vernier caliper of 0.02mm
resolution (The uncertainties of the measurements are probably larger
than 0.1mm I think). The result are compared to hypothesis values. The
maximum error is around 1mm (about 3~4 pixels in our application).

distance | hypothesis | measured
---------|------------|---------
\|p1-p2\|| 225        | 224.14  
\|p1-p3\|| 135        | 135.35  
\|p1-p4\|| 262.393    | 261.85  
\|p2-p3\|| 262.393    | 262     
\|p2-p4\|| 135        | 135.13  
\|p3-p4\|| 225        | 224.73  

The inaccuracy of the checkerboard causes less precise and less stable
estimation of camera intrinsic parameters.

### New method integrated

This PR integrates an extension to the standard pinhole camera
calibration method in OpenCV. The method was proposed in the paper:
Klaus H. Strobl and Gerd Hirzinger. [More accurate pinhole camera
calibration with imperfect planar target](https://elib.dlr.de/71888/1/strobl_2011iccv.pdf). In 2011 IEEE International
Conference on Computer Vision (ICCV), pages 1068-1075, Barcelona, Spain,
Nov 2011. IEEE. And it was included in the camera calibration toolbox
[DLR CalDe and DLR CalLab](http://www.robotic.dlr.de/callab/).
In many common cases with inaccurate calibration plates, this method can
dramatically improve the precision of the estimated camera parameters.

In principle the method releases 3D coordinates of the object structure
when performing non-linear Levenberg-Marquardt optimization. The current
calibration process can be easily reinforced by adding extra
optimization parameters and extending the Jacobian matrix.

Before trying to integrate it into OpenCV, a standalone project
[calibrel](https://github.com/xoox/calibrel) was created to reimplement
and validate this method. [The results were double-checked](https://github.com/xoox/calibrel/issues/1#issuecomment-419775155) with [CalLab](http://www.robotic.dlr.de/callab/).
In addition to validating with stereo cameras in the paper, [we also
evaluated its precision using measured distances](https://github.com/xoox/calibrel/issues/2).

When preparing to add test code for this new method, [the existing data
for regression test is reordered](https://github.com/xoox/opencv_extra/blob/calib-release-object/testdata/cv/cameracalibration/calib1.dat) where the object points and image
feature points can be also used by the new calibration method (with
calibration flags changed). The results with releasing object points are
so diverse to standard methods. It may be caused by bad set of images or
inaccuracy of feature points. The comparison is shown in the following
table and further discussions are needed. Could anyone do some
comparisons between these two methods with regard to the same set of
images of an **_accurate_** checkerboard? Both cameras with larger and
smaller distortions are needed. Experiments can be conducted with
`samples/cpp/calibration.cpp` and
`samples/cpp/tutorial_code/calib3d/camera_calibration/camera_calibration.cpp`
in this PR.

parameter | standard   | release_object
----------|------------|---------------
fx        | 809.94872  | 790.93722
fy        | 806.56095  | 775.18624
cx        | 361.99611  | 331.43434
cy        | 214.39525  | 203.89052
k1        | -0.1282055 | -0.244456
k2        | -0.3944258 | 0.3578029
p1        | -0.0043549 | -0.002096
p2        | -0.0073207 | -0.001801
RMS       | 0.138892   | 0.0830433

I have no information of the camera, lens and checkerboard used in
`calib1.dat`. But after drawing its feature points on images, it seems those
images are probably not good candidates for camera calibration. Please
check the images on this PR opencv/opencv_extra#528. Also see [Dr.  Strobl](https://rmc.dlr.de/rm/en/staff/klaus.strobl/)'s
comments on good calibration images: https://github.com/xoox/calibrel/issues/1#issuecomment-419279616

If this PR is going to be squashed, it'd be better full commit messages
are kept instead of commit titles only, for that there are useful
information in the full commit messages.

In the similar fashion, this method can also be included into
`stereoCalibrate()`. But the speed is expected much slower. If the
intrinsic parameters can be estimated with high accuracy for each of the
cameras individually using `calibrateCamera()`, it is recommended to do
so and then pass CALIB_FIX_INTRINSIC flag to `stereoCalibrate()` along
with the computed intrinsic parameters. However, I'd integrate this
method into `stereoCalibrate()` if the necessity is confirmed.

#### cvCalibrateCamera4()

A new C interface is added which supports both standard calibration
method and the extended new calibration method.

#### calibrateCameraRO()

New C++ interfaces are added which support both standard calibration
method and the extended new calibration method.

#### Regression test code

The test code is modified to empower tests of both standard calibration
method and the new added one. The updated test data can be found here:
https://github.com/xoox/opencv_extra/tree/calib-release-object A PR opencv/opencv_extra#528 was
also created for the new test data.

#### Sample calibration code

Two calibration samples are updated to support both standard
calibration method and the extended new calibration method.

In conclusion, this PR is just a very straightforward extension to
current standard calibration code.

### CODE REVIEW

**Comment by xoox** [comment]
The build failed because modified test data are needed. The data can be found here: opencv/opencv_extra#528
Is there any way to control the build bot to use a specific branch of opencv_extra?

File: modules/calib3d/include/opencv2/calib3d.hpp [hunk 1]
```diff
@@ -1080,23 +1101,38 @@ The algorithm performs the following steps:
    findChessboardCorners, solvePnP, initCameraMatrix2D, stereoCalibrate, undistort
  */
 CV_EXPORTS_AS(calibrateCameraExtended) double calibrateCamera( InputArrayOfArrays objectPoints,
-                                     InputArrayOfArrays imagePoints, Size imageSize,
+                                     InputArrayOfArrays imagePoints, Size imageSize, int iFixedPoint,
```
Comment (alalek) [inline modules/calib3d/include/opencv2/calib3d.hpp]: Lets preserve existing interfaces and add new one (something like `calibrateCameraReleaseObject`).

> iFixedPoint

Looks like, this and other new parameters imply `flags += CALIB_RELEASE_OBJECT` automatically.

Comment (xoox) [reply modules/calib3d/include/opencv2/calib3d.hpp]: > Lets preserve existing interfaces and add new one (something like `calibrateCameraReleaseObject`).

That's my initial thought before starting. With regard to the fact that
they share most of the interface documents, I chose functions overload.
OK, I'll add new interfaces instead of overload old ones. How about the
name `calibrateCameraRO`? It's shorter and attractive to users, but with
the meaning less obvious. Since most part of the interface document are
shared by `calibrateCamera` and `calibrateCameraRO`,
`calibrateCameraRO`'s interface document will be simpler with most
explanations referenced to `calibrateCamera`.

The standard calibration method and object-releasing calibration method
will be implemented in `calibrateCameraRO`, and method is chosen based
on whether `iFixedPoint` is within a rational range. `calibrateCamera`
is then made a wrapper to `calibrateCameraRO`.

> 
> > iFixedPoint
> 
> Looks like, this and other new parameters imply `flags += CALIB_RELEASE_OBJECT` automatically.

Only `iFixedPoint` is input parameter, the other two are optional output
parameters. You are right. The method selection can be controlled with
`iFixedPoint`. I'll remove `CALIB_RELEASE_OBJECT` item and make the
method selection based on the value of `iFixedPoint`.

File: modules/calib3d/src/calibration.cpp [hunk 2]
```diff
@@ -1716,8 +1862,47 @@ CV_IMPL double cvCalibrateCamera2( const CvMat* objectPoints,
...
+                    CvMat* rvecs, CvMat* tvecs, CvMat* newObjPoints, int flags, CvTermCriteria termCrit )
+{
+    // If iFixedPoint is out of rational range, fall back to standard method
+    if( iFixedPoint > npoints->data.i[0] - 2 )
+        iFixedPoint = -1;
+    // check object points. If not qualified, fall back to standard calibration.
+    int nimages = npoints->rows * npoints->cols;
+    int npstep = npoints->rows == 1 ? 1 : npoints->step / CV_ELEM_SIZE(npoints->type);
+    int i, ni;
+    if( iFixedPoint > 0 )
```
Comment (alalek) [inline modules/calib3d/src/calibration.cpp]: Why `iFixedPoint == 0` is out of scope? Is it allowed?

> The index of the 3D object point in objectPoints[0] to be fixed. Usually it is
the top-right corner point of the calibration board grid.

Index is zero-based in C++. Zero index should correspond to "top-left" corner.

Comment (xoox) [reply modules/calib3d/src/calibration.cpp]: Based on the cited paper, three 3D object points are chosen to be fixed. All x, y, z coordinates of point1 and point2 are fixed, and only z of point3 is fixed. In theory, these points could be arbitrarily chosen as long as they are non-collinear. but in practice, it will deliver more accurate results to select points that are far away from each other. In the internal implementation, the first (zero index, "top-left" corner) point is already selected as point1, see line 1651-1653. And the z coordinate of the last point ("bottom-right" corner) is also fixed as point3.z, see line 1657. Users have the freedom to select point2, and "top-right" or "bottom-left" point of the calibration board grid is recommended.

Use the convention in the sample C++ code, the first point ("top-left" corner) has the coordinates (0, 0, 0), "top-right" corner has (d, 0, 0) and the last point ("bottom-right" corner) has (x3, y3, 0). The accuracy of intrinsic parameters is free from these fixed points. But for accurate rvecs, tvecs and newObjPoints, the fixed coordinates must be accurate enough. In the above convention, only "d" should be accurately determined. "d" can be easily measured as the distance between the "top-left" and "top-right" corners.

There're samples on how to set the accurate "d". `samples/cpp/calibration.cpp` line 162 and `samples/cpp/tutorial_code/calib3d/camera_calibration/camera_calibration.cpp` line 561.

Comment (alalek) [reply modules/calib3d/src/calibration.cpp]: Thank you for detailed explanation.

Perhaps we should have documentation notes about this (specify valid / expected range of input values):
> In the internal implementation, the first (zero index, "top-left" corner) point is already selected as point1

> And the z coordinate of the last point ("bottom-right" corner) is also fixed as point3.z

Also add parameters checks into code like this:
```
CV_Assert(iFixedPoint != 0);
CV_Assert(iFixedPoint != numpoints - 1);
```
(not sure about second one)

Comment (xoox) [reply modules/calib3d/src/calibration.cpp]: The current document for `iFixedPoint` has the following information:

```
...
According to \cite strobl2011iccv, two other points are also fixed. In
this implementation, objectPoints[0].front and objectPoints[0].back.z
are used. Accurate rvecs and tvecs are only possible if coordinates of
these three fixed points are accurate enough.
...
```

I'd like to update it with the following text:

```
@param iFixedPoint The index of the 3D object point in objectPoints[0] to be fixed. Usually
the top-right corner point of the calibration board grid is recommended. For values out of the range
[1, objectPoints[0].size()-2], calibration will fall back to standard method. According to
\cite strobl2011iccv, two other points are also fixed. In this implementation, objectPoints[0].front
and objectPoints[0].back.z are used. Accurate rvecs, tvecs and newObjPoints are only possible if
coordinates of these three fixed points are accurate enough. If standard calibration method to be
used, just set this parameter to a negative value.
```

Is it acceptable?

------------------------------------------------------------------------

I think soft limit is a better experience for users, i.e. just falling back
to standard calibration method if `iFixedPoint < 1` or `iFixedPoint > numpoints - 2`.
If accuracy isn't the most important thing for some users, then fallback
scheme is acceptable for them. It seems assertions are a little bit
harsh for this parameter.

Comment (alalek) [reply modules/calib3d/src/calibration.cpp]: Before using, users would like to compare accuracy between old and new functions. What parameters are fit for their tasks.

I believe we should not confuse them with the same results (and silently disabled algorithm) if `iFixedPoint` is out of valid range for this algorithm (error message is a normal way to say that input values is out of expected range - you may use `CV_Check()` macros to add more accurate error message).

Comment (catree) [reply modules/calib3d/src/calibration.cpp]: I also agree that silently falling back to the classical `calibrateCamera` is not a good thing.

Another [example](https://github.com/opencv/opencv/issues/5065) that led users to report issues are the `SOLVEPNP_DLS` and `SOLVEPNP_UPNP` flags in `SolvePnP` function.

Comment (xoox) [reply modules/calib3d/src/calibration.cpp]: OK, the confusing might arise from the fact that we are using
`iFixedPoint` as a switch for calibration method selection. It'd be
better to not use "fall back" words. Just tell users to select new
method, pass in the parameter in the specified range. If standard method
is demanded, set it out of the range.

If new method is requested, but the `objectPoints` data are not qualified,
then `CV_Assert()` or `CV_error()` is called.

Thanks for your comments. I'll add more commits for further review.

File: doc/tutorials/calib3d/table_of_content_calib3d.markdown [hunk 3]
```diff
@@ -13,7 +13,7 @@ Although we get most of our images in a 2D format they do come from a 3D world.
 
 -   @subpage tutorial_camera_calibration
 
-    *Compatibility:* \> OpenCV 2.0
+    *Compatibility:* \> OpenCV 4.1
```
Comment (xoox) [inline doc/tutorials/calib3d/table_of_content_calib3d.markdown]: If this feature could be released with 4.0, the compatibility would be updated to 4.0.

Comment (xoox) [inline doc/tutorials/calib3d/table_of_content_calib3d.markdown]: If this feature could be released with 4.0, the compatibility would be updated to 4.0.

**Comment by alalek** [comment]
@xoox Well done!

Please update tutorial version to 4.0.

Squashed commit will look like this: https://github.com/alalek/opencv/commit/pr12772_r (if you want another commit message please add comment with updated version).
